cli/deno_dir.rs

```diff
@@ -445,49 +533,135 @@ fn get_source_code(
+static VERSION_HASH: &'static str = "version_hash";
+
+impl CompiledFileMetadata {
+  /// Get compiled file metadata from meta_path.
+  /// If operation failed or metadata file is corrupted,
+  /// return None.
+  pub fn load<P: AsRef<Path>>(meta_path: P) -> Option<CompiledFileMetadata> {
+    let meta_path = meta_path.as_ref();
+    let maybe_metadata_string = fs::read_to_string(meta_path).ok();
+    maybe_metadata_string.as_ref()?;
+    let maybe_metadata_json: serde_json::Result<serde_json::Value> =
+      serde_json::from_str(&maybe_metadata_string.unwrap());
+    if let Ok(metadata_json) = maybe_metadata_json {
+      let source_path = metadata_json[SOURCE_PATH].as_str().map(String::from);
+      let version_hash = metadata_json[VERSION_HASH].as_str().map(String::from);
+      if source_path.is_none() || version_hash.is_none() {
+        return None;
+      }
+      return Some(CompiledFileMetadata {
+        source_path: source_path.unwrap(),
+        version_hash: version_hash.unwrap(),
+      });
+    } else {
+      return None;
+    }
+  }
+
+  /// Save compiled file metadata to meta_path.
+  pub fn save<P: AsRef<Path>>(self: &Self, meta_path: P) {
+    let meta_path = meta_path.as_ref();
+    // Remove possibly existing stale meta file.
+    // May not exist. DON'T unwrap.
+    let _ = std::fs::remove_file(&meta_path);
+    let mut value_map = serde_json::map::Map::new();
+
+    value_map.insert(SOURCE_PATH.to_owned(), json!(&self.source_path));
+    value_map.insert(VERSION_HASH.to_string(), json!(&self.version_hash));
+
+    let _ = serde_json::to_string(&value_map).map(|s| {
+      let _ = deno_fs::write_file(meta_path, s, 0o666);
+    });
+  }
+}
+
+/// Try loading the compiled code and map from the cache.
+/// Also validate the state hash in case source code has changed.
+/// If the version hash does not match,
+/// try delete all these files and return an error.
+fn load_cache(
   js_filename: &PathBuf,
   map_filename: &PathBuf,
+  meta_filename: &PathBuf,
+  version_hash_to_validate: String,
 ) -> Result<(Vec<u8>, Vec<u8>), std::io::Error> {
   debug!(
-    "load_cache code: {} map: {}",
+    "load_cache code: {} map: {} meta: {}",
     js_filename.display(),
-    map_filename.display()
+    map_filename.display(),
+    meta_filename.display()
   );
   let read_output_code = fs::read(&js_filename)?;
   let read_source_map = fs::read(&map_filename)?;
+
+  // Validate metadata.
+  let mut should_remove_all = false;
+  if let Some(read_metadata) = CompiledFileMetadata::load(meta_filename) {
+    if read_metadata.version_hash != version_hash_to_validate {
+      debug!("load_cache metadata version hash does not match");
+      should_remove_all = true;
+    }
+  } else {
+    debug!("load_cache metadata load error");
+    should_remove_all = true;
+  }
+
+  // TODO: this step can be skipped now
+  // The old version is stale or some error occurred.
+  // Try removing all the cached files.
+  if should_remove_all {
```

Comment:
Since filenames are no longer hashes there'll be only file generated file per one input file:
```
https://deno.land/std/http/file_server.ts
// path in DENO_DIR to compiled module
$DENO_DIR/gen/https/deno.land/std/http/file_server.js
$DENO_DIR/gen/https/deno.land/std/http/file_server.js.map
$DENO_DIR/gen/https/deno.land/std/http/file_server.meta
```

 I believe this step can be skipped now.